torchrl/data/replay_buffers/replay_buffers.py

```diff
@@ -858,67 +953,48 @@ def add(self, data: TensorDictBase) -> int:
-
-        tensordicts = TensorDict(
-            {"_data": tensordicts},
-            batch_size=tensordicts.batch_size[:1],
-        )
-        if tensordicts.batch_dims > 1:
-            # we want the tensordict to have one dimension only. The batch size
-            # of the sampled tensordicts can be changed thereafter
-            if not isinstance(tensordicts, LazyStackedTensorDict):
-                tensordicts = tensordicts.clone(recurse=False)
-            else:
-                tensordicts = tensordicts.contiguous()
-            # we keep track of the batch size to reinstantiate it when sampling
-            if "_rb_batch_size" in tensordicts.keys():
-                raise KeyError(
-                    "conflicting key '_rb_batch_size'. Consider removing from data."
-                )
-            shape = torch.tensor(tensordicts.batch_size[1:]).expand(
-                tensordicts.batch_size[0], tensordicts.batch_dims - 1
+        if not isinstance(tensordicts, TensorDictBase):
+            raise ValueError(
+                f"{self.__class__.__name__} only accepts TensorDictBase subclasses. tensorclasses "
+                f"and other types are not compatible with that class. "
+                "Please use a regular `ReplayBuffer` instead."
             )
-            tensordicts.set("_rb_batch_size", shape)
-        tensordicts.set(
-            "index",
-            torch.zeros(tensordicts.shape, device=tensordicts.device, dtype=torch.int),
-        )
-
         if self._transform is not None:
-            data = self._transform.inv(tensordicts.get("_data"))
-            tensordicts.set("_data", data)
-            if data.device is not None:
-                tensordicts = tensordicts.to(data.device)
+            tensordicts = self._transform.inv(tensordicts)
 
         index = super()._extend(tensordicts)
+        self._set_index_in_td(tensordicts, index)
         self.update_tensordict_priority(tensordicts)
         return index
 
+    def _set_index_in_td(self, tensordict, index):
+        if index is None:
+            return
+        if _is_int(index):
+            index = torch.as_tensor(index, device=tensordict.device)
+        elif index.ndim == 2 and index.shape[:1] != tensordict.shape[:1]:
+            for dim in range(2, tensordict.ndim + 1):
+                if index.shape[:1].numel() == tensordict.shape[:dim].numel():
+                    # if index has 2 dims and is in a non-zero format
+                    index = index.unflatten(0, tensordict.shape[:dim])
+                    break
+            else:
+                raise RuntimeError(
+                    f"could not find how to reshape index with shape {index.shape} to fit in tensordict with shape {tensordict.shape}"
+                )
+            tensordict.set("index", index)
+            return
+        tensordict.set("index", expand_as_right(index, tensordict))
+
     def update_tensordict_priority(self, data: TensorDictBase) -> None:
```

Comment:
The signature of the method has changed since we added `return self.update_priority(index, priority)`